test/end2end_test.go

```diff
@@ -6880,6 +6881,174 @@ func (s *httpServer) start(t *testing.T, lis net.Listener) {
+	authInfo := &authInfoWithConn{CommonAuthInfo: credentials.CommonAuthInfo{SecurityLevel: credentials.NoSecurity}, connection: conn}
+	return conn, authInfo, nil
+}
+func (c *transportRestartCheckCreds) Info() credentials.ProtocolInfo {
+	return credentials.ProtocolInfo{}
+}
+func (c *transportRestartCheckCreds) Clone() credentials.TransportCredentials {
+	return &transportRestartCheckCreds{}
+}
+func (c *transportRestartCheckCreds) OverrideServerName(s string) error {
+	return nil
+}
+
+type streamWithPeer struct {
+	stream testpb.TestService_FullDuplexCallClient
+	peer   *peer.Peer
+	id     string
+	conn   net.Conn
+}
+
+func (s *streamWithPeer) sendAndReceive() error {
+	if err := s.stream.Send(&testpb.StreamingOutputCallRequest{}); err != nil {
+		return fmt.Errorf("sending on %s: %v", s.id, err)
+	}
+	if _, err := s.stream.Recv(); err != nil {
+		return fmt.Errorf("receiving on %s: %v", s.id, err)
+	}
+	return nil
+}
+
+// Tests that the client transport drains and restarts when next stream ID exceeds
+// MaxStreamID. This test also verifies that subsequent RPCs use a new client
+// transport and the old transport is closed.
+func (s) TestClientTransportRestartsAfterStreamIDExhausted(t *testing.T) {
+	// Set the transport's MaxStreamID to 5 to cause connection to drain after 2 RPCs.
+	originalMaxStreamID := transport.MaxStreamID
+	transport.MaxStreamID = 5
+	defer func() {
+		transport.MaxStreamID = originalMaxStreamID
+	}()
+
+	ss := &stubserver.StubServer{
+		FullDuplexCallF: func(stream testpb.TestService_FullDuplexCallServer) error {
+			for {
+				_, err := stream.Recv()
+				if err == io.EOF {
+					return nil
+				}
+				if err != nil {
+					return err
+				}
+				if err := stream.Send(&testpb.StreamingOutputCallResponse{}); err != nil {
+					return err
+				}
+			}
+		},
+	}
+
+	creds := &transportRestartCheckCreds{}
+	if err := ss.Start(nil, grpc.WithTransportCredentials(creds)); err != nil {
+		t.Fatalf("Starting stubServer: %v", err)
+	}
+	defer ss.Stop()
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	var streams []streamWithPeer
+
+	// Setting up 3 streams and calling Send() and Recv() once on each.
+	for i := 0; i < 3; i++ {
+		p := new(peer.Peer)
+		s, err := ss.Client.FullDuplexCall(ctx, grpc.Peer(p))
+		if err != nil {
+			t.Fatalf("Creating FullDuplex stream: %v", err)
+		}
+		sp := streamWithPeer{stream: s, peer: p, id: "stream " + strconv.Itoa(i)}
+		streams = append(streams, sp)
+
+		if err := sp.sendAndReceive(); err != nil {
+			t.Fatal(err)
+		}
+	}
+	// Verifying only 2 connections are created so far.
+	if len(creds.connections) != 2 {
+		t.Fatalf("Number of connections created: %v, want: 2", len(creds.connections))
+	}
+
+	// Verifying that streams on the first conn still works.
+	if err := streams[0].sendAndReceive(); err != nil {
+		t.Fatal(err)
+	}
+	if err := streams[1].sendAndReceive(); err != nil {
+		t.Fatal(err)
+	}
```

Comment:
Just do all three streams again for simplicity and consistency?

Reply:
I had to make it for 3 streams anyways, when i updated the FullDuplexCallF func